Switch from iconv-lite to exodus/bytes for decoding

ChALkeR opened the pull request:
See https://github.com/jsdom/whatwg-encoding/issues/22 and https://github.com/jsdom/whatwg-encoding/pull/24#issuecomment-3672970117

This should be non-breaking:
1. `legacyHookDecode` now accepts everything that `whatwg-encoding.decode()` accepts (as it supports all labels)
2. Added `labelToName` that returns cased names for the DOM API
3. Engines support remains the same

This fixes handling of all encodings that were previously incorrect, including: `UTF-8`, `UTF-16`, half of single-byte, all of multi-byte, and adds `ISO-8859-8-I`, `x-mac-cyrillic`, `ISO-2022-JP`, `replacement`

Calls to `legacyHookDecode` don't check that the encoding is not `'replacement'`. Should they?

domenic commented:
I did the new v5 html-encoding-sniffer release so we can roll that into this too.

> 1. `labelToName` usage blocks some encodings that are actually supported

Can you expand on what you mean by this? What was the previous behavior, and what is the new behavior?

> 2\. Calls to `legacyHookDecode` don't check that the encoding is not `'replacement'`. Should they?

I suspect this is fine as-is, as there is no such check in the spec. For example, if replacement shows up in the charset= parameter of a Content-Type for an iframe, style, or script, I suspect this is supposed to work.

domenic commented on lib/jsdom/living/nodes/HTMLScriptElement-impl.js:
```diff
@@ -88,10 +88,10 @@ class HTMLScriptElementImpl extends HTMLElementImpl {
 
-      const encoding = whatwgEncoding.getBOMEncoding(data) ||
-        (contentType && whatwgEncoding.labelToName(contentType.parameters.get("charset"))) ||
+      const encoding = getBOMEncoding(data) ||
```
Rename to `canonicalizedEncodingLabel`

domenic commented on lib/jsdom/living/xhr/XMLHttpRequest-impl.js:
```diff
@@ -301,7 +302,7 @@ class XMLHttpRequestImpl extends XMLHttpRequestEventTargetImpl {
         parsingMode: isHTML ? "html" : "xml",
         cookieJar: { setCookieSync: () => undefined, getCookieStringSync: () => "" },
-        encoding,
+        encoding: toCasedEncodingName(encoding),
```
Rename the constructor privateData option to `Document` to `canonicalizedEncodingLabel` (which sets `_canonicalizedEncodingLabel`), and then we can just pass it in as-is.

ChALkeR replied:
This was reverted

domenic commented on lib/api.js:
```diff
@@ -301,7 +302,8 @@ function normalizeHTML(html, mimeType) {
     });
-    html = whatwgEncoding.decode(html, encoding);
+    html = legacyHookDecode(html, encoding);
+    encoding = toCasedEncodingName(encoding);
```
Remove this since per previous discussion we'll store the canonicalized encoding labels.

ChALkeR commented:
> I don't think this can be removed, because the spec requires exposing the encoding name:

Huh, I see. I missed that those names are exposed in string form in DOM API, not just the internal structure.
I might rethink that

domenic commented:
Yeah, I learned today for the first time that the getters on `Document` return different values than the ones on `TextDecoder` and `TextDecoderStream`. The web is so wonderful... 🙃

ChALkeR commented:
I added `labelToName` and wrapped most calls to `getBOMEncoding` and `sniffHTMLEncoding` in it
Except one case where label can be used directly, there I renamed the var.

`normalizeEncoding` is not used anymore.

ChALkeR commented on lib/jsdom/living/file-api/FileReader-impl.js:
```diff
@@ -40,7 +40,7 @@ class FileReaderImpl extends EventTargetImpl {
     this._readFile(file, "dataURL");
   }
   readAsText(file, encoding) {
```
Done

ChALkeR commented on lib/jsdom/living/file-api/FileReader-impl.js:
Renamed to `encodingLabel`, does not have to be canonical.

ChALkeR commented on lib/jsdom/living/nodes/HTMLFrameElement-impl.js:
```diff
@@ -51,10 +52,10 @@ function fetchFrame(serializedURL, frame, document, contentDoc) {
     }
 
-    const encoding = sniffHTMLEncoding(data, sniffOptions);
+    const encoding = toCasedEncodingName(sniffHTMLEncoding(data, sniffOptions));
```
After https://github.com/jsdom/jsdom/pull/4004#issuecomment-3686023238, I just added `labelToName` and kept the old approach which stores case-sensitive names everywhere (which seems to be what DOM API wants).

`legacyHookDecode` can accept any valid label now and doesn't need them to be normalized, so passing names is fine.

domenic commented on lib/jsdom/living/file-api/FileReader-impl.js:
```diff
@@ -39,8 +39,8 @@ class FileReaderImpl extends EventTargetImpl {
-  readAsText(file, encoding) {
-    this._readFile(file, "text", whatwgEncoding.labelToName(encoding) || "UTF-8");
+  readAsText(file, encodingLabel) {
+    this._readFile(file, "text", labelToName(encodingLabel) || "UTF-8");
```
We can remove this wrapping now then I guess?

ChALkeR replied:
This is different.


`legacyHookDecode` accepts _any valid label_ and throws otherwise.
`labelToName` maps unrecognized labels to `null`

So, what this code does is maps any unrecognized labels to `UTF-8`.
This matches the old behavior, changing the behavior on unrecognized labels would be out of scope of this PR.

```js
> require('whatwg-encoding').decode(Uint8Array.of(), 'what')
Uncaught RangeError: "what" is not a supported encoding name
> encoding = require('whatwg-encoding').labelToName('what') || 'UTF-8'
> require('whatwg-encoding').decode(Uint8Array.of(), encoding)
''

> require('@exodus/bytes/encoding.js').legacyHookDecode(Uint8Array.of(), 'what')
Uncaught RangeError: Unknown encoding
> encoding = require('@exodus/bytes/encoding.js').labelToName('what') || 'UTF-8'
> require('@exodus/bytes/encoding.js').legacyHookDecode(Uint8Array.of(), encoding)
''

ChALkeR commented on lib/jsdom/living/helpers/stylesheets.js:
`document._encoding` is case-sensitive name and that wasn't changed

ChALkeR commented on lib/jsdom/living/xhr/XMLHttpRequest-impl.js:
```diff
@@ -289,8 +290,8 @@ class XMLHttpRequestImpl extends XMLHttpRequestEventTargetImpl {
 
-    const encoding = finalCharset(this) || whatwgEncoding.getBOMEncoding(responseBuffer) || "UTF-8";
-    const resText = whatwgEncoding.decode(responseBuffer, encoding);
+    const encoding = finalCharset(this) || getBOMEncoding(responseBuffer) || "utf-8";
```
This one is back to being a case-sensitive name.

domenic commented on lib/jsdom/living/file-api/FileReader-impl.js:
```diff
@@ -39,8 +39,8 @@ class FileReaderImpl extends EventTargetImpl {
-  readAsText(file, encoding) {
-    this._readFile(file, "text", whatwgEncoding.labelToName(encoding) || "UTF-8");
+  readAsText(file, encodingLabel) {
+    this._readFile(file, "text", labelToName(encodingLabel) || "UTF-8");
```
```suggestion
    this._readFile(file, "text", encodingLabel || "utf-8");
```

domenic commented on lib/jsdom/living/nodes/HTMLScriptElement-impl.js:
```diff
@@ -62,7 +62,7 @@ class HTMLScriptElementImpl extends HTMLElementImpl {
     const document = this._ownerDocument;
     const resourceLoader = document._resourceLoader;
-    const defaultEncoding = whatwgEncoding.labelToName(this.getAttributeNS(null, "charset")) || document._encoding;
+    const defaultEncoding = labelToName(this.getAttributeNS(null, "charset")) || document._encoding;
```
We can change this to defaultEncodingLabel and remove labelToName, I think?

ChALkeR commented on lib/jsdom/encoding-map.js:
```diff
@@ -0,0 +1,20 @@
+// Map lowercase @exodus/bytes/encoding@1 encoding names to cased whatwg-encoding@3 names
+// Does not perform any input validation, only output from @exodus/bytes/encoding or html-encoding-sniffer is expected
+// null falls through
+// This logic is only for compat and can be removed by switching exposed encoding properties to lowercase
```
Based on https://github.com/jsdom/jsdom/pull/4004#issuecomment-3686023238, I just added `labelToName` to `@exodus/bytes`, which solves this.

domenic reviewed (commented):
Alright, so now `@exodus/bytes`'s `legacyDecodeHook` accepts all labels, and labels are a superset of names. So:

- We can continue storing the names as `_encoding` on `Document` objects.
- We can otherwise just remove almost all canonicalization code, operating on raw labels we get from e.g. `charset=""` attributes or `Content-Type` headers or whatever, since it doesn't really matter what we pass in to the underlying package.

I think my latest round of suggestions/comments gets us to that state. Let me know what you think.

This seems like a pretty nice architecture in the end.

domenic reviewed (approved):
Thanks for walking me through this!
